packages/backend/src/dbt/dbtRpcClientBase.ts

```diff
@@ -0,0 +1,291 @@
+import { v4 as uuidv4 } from 'uuid';
+import fetch from 'node-fetch';
+import {
+    DbtRpcCompileResults,
+    DbtRpcDocsGenerateResults,
+    isDbtRpcCompileResults,
+    isDbtRpcDocsGenerateResults,
+    isDbtRpcRunSqlResults,
+} from 'common';
+import { DbtError, NetworkError, RetryableNetworkError } from '../errors';
+
+export const DEFAULT_HEADERS: Record<string, string> = {
+    'Content-Type': 'application/json',
+};
+
+type PollArgs<T> = {
+    func: () => Promise<T>;
+    condition: (t: T) => Promise<boolean>;
+    maxAttempts: number;
+    startInterval: number;
+    maxInterval: number;
+    intervalMultiplier: number;
+};
+
+const pollOverNetwork = async <T>({
+    func,
+    condition,
+    maxAttempts,
+    startInterval,
+    maxInterval,
+    intervalMultiplier,
+}: PollArgs<T>): Promise<T> => {
+    let attempts = 0;
+    let interval = startInterval;
+
+    const poll = async (
+        resolve: (res: T) => void,
+        reject: (err: any) => void,
+    ) => {
+        if (attempts >= maxAttempts) {
+            reject(
+                new NetworkError(
+                    `Lightdash timedout trying to reach dbt server.`,
+                    {},
+                ),
+            );
+            return;
+        }
+        try {
+            const result = await func();
+            const success = await condition(result);
+            if (success) {
+                resolve(result);
+                return;
+            }
+        } catch (e) {
+            if (!(e instanceof RetryableNetworkError)) {
+                reject(e);
+                return;
+            }
+        }
+        attempts += 1;
+        interval = Math.min(maxInterval, interval * intervalMultiplier);
+        setTimeout(poll, interval, resolve, reject);
```

Comment:
We should extract the retry mechanism into a util function. 
I found this answer and it looks clean, what you think ? 
https://stackoverflow.com/a/44577075

Reply:
I thought this was pretty barebones. It's reusable retry that works with Lightdash specific Errors. 

What two functions would you break this into?

Reply:
It was in case we wanted to use a generic promise retry in multiple places beside connecting to dbt server. 
We don't have to change this now and we can re-evaluate later if we need to.